Frontend for subscription

NBNARADHYA opened the pull request:
# Changes/Additions
1. Attempt to fetch `creatorAccount` and `subscriptions` for the user.
2. Button to reroute to `creator-hub` or `subscription-hub`.
3. `User` context and `useUser` hook to get user details in all pages.
4. `isSubscribed` map in `User` context to see if a user is subscribed to a creator or not. 

Usage for point 4.
```
if (isSubscribed[creator.pubKey.toString()]) {
   // The user is subscribed to the creator
} else {
   // The user is not subscribed to the creator.
}

MikaelCarpenter commented on app/pages/index.tsx:
```diff
@@ -98,12 +145,32 @@ const Home: NextPage = () => {
+            >
+              {user.subscriptions.length
+                ? 'See my subsriptions'
+                : 'Subscribe to a creator'}
```
For this we can ignore the case if the user has no subscriptions. So if they don't have subscriptions they just see the one button for "Become a Creator"

NBNARADHYA replied:
The user should be able to navigate right ? to their subscriptions ?

NBNARADHYA replied:
Or should we put those things in a menu item ?

MikaelCarpenter replied:
Correct, we'll have a page where a user can see all of their subscriptions.

NBNARADHYA replied:
yes I believe I am creating that page -> subsription-hub. I meant to say how would they access that page ? Like a navigation

MikaelCarpenter replied:
The landing page should have a button for that, which is this one for "See my subscriptions". And the CreatorHub page will have a link to it as well, in case a creator wants to get there

MikaelCarpenter commented on app/pages/subscription-hub.tsx:
```diff
@@ -0,0 +1,24 @@
+  user.creatorAccount && router.push('/creator-hub');
+
+  // Reroute to `/subscribe` if user hasn't subscribed to any content.
+  !user.subscriptions.length && router.push('/subscribe');
```
in this case we can reroute them to the landing page. we won't have a `/subscribe` route, they'll have to go directly to a creator landing page to subscribe to a creator

NBNARADHYA replied:
I see. I thought creator landing page was specific to particular creator. I meant that only by `/subscribe`.

MikaelCarpenter replied:
Correct, each creator gets their own creator landing page

NBNARADHYA replied:
By `/subscribe` I meant a page which would show different creators. IG we can call that `/creators` ? And reroute the user to that page ?

MikaelCarpenter replied:
yea totally 👍  that sounds good

MikaelCarpenter commented on app/pages/subscription-hub.tsx:
```diff
@@ -0,0 +1,24 @@
+  !connectedWallet && router.push('/');
+
+  // Reroute to `/creator-hub` if the user is a creator.
+  user.creatorAccount && router.push('/creator-hub');
```
A creator may also be a subscriber, so they'll want to be able to look at this page that shows all of their subscriptions

NBNARADHYA replied:
Yes makes sense

MikaelCarpenter commented on app/pages/index.tsx:
```diff
@@ -27,41 +28,87 @@ const Home: NextPage = () => {
+        setUser({
+          subscriptions,
+          creatorAccount: creator,
+        });
```
It may be worth moving all of this logic up a level to `_app.tsx` because no matter what route a user first hits, we'll want to fetch this kind of data

MikaelCarpenter reviewed (commented):
This is looking really good!! 👍

NBNARADHYA commented:
1. Moved all fetching code back to `index.tsx` because we have to make sure `WalletProvider` is wrapped in an ancestor component.
2. Extended `fetchSubscriptionsForUserWallet` function to fetch benefits for the creators which the user has subscribed as well.

NBNARADHYA commented:
@MikaelCarpenter I have removed subscription-hub and creator. pages. The fetching logic  remains,. I have also added `isSubscribed` map to check if a user is subscribed to a particular creator.

NBNARADHYA commented:
I am creating a new pr for the `subscription-hub` and `creators` pages.

MikaelCarpenter commented on app/hooks/userUser.ts:
```diff
@@ -0,0 +1,16 @@
+import { createContext, useContext } from 'react';
+
+export interface User {
+  creatorAccount: Record<string, unknown> | null;
```
not needed for this to merge, but we should define actual types for each of these. @KitanGarcia may have already started that, so we can try to stitch that together here

NBNARADHYA replied:
Sure!

MikaelCarpenter reviewed (approved):
looks great! 👍
